File: LEM_Listas/Listas_Enlazadas.cs
using System;
using System.Collections.Generic;

public class Nodo
{
    public int data;
    public Nodo next;

    public Nodo(int data)
    {
        this.data = data;
        this.next = null;
    }
}

public class ListaEnlazada
{
    private Nodo head;

    public ListaEnlazada()
    {
        head = null;
    }

    public void InsertarAlPrincipio(int data)
    {
        Nodo nuevo = new Nodo(data);
        nuevo.next = head;
        head = nuevo;
    }

    public void InsertarAlFinal(int data)
    {
        Nodo nuevo = new Nodo(data);
        if (head == null)
        {
            head = nuevo;
            return;
        }
        Nodo ultimo = head;
        while (ultimo.next != null)
        {
            ultimo = ultimo.next;
        }
        ultimo.next = nuevo;
    }

    public void ImprimirLista()
    {
        Nodo actual = head;
        while (actual != null)
        {
            Console.Write(actual.data + " -> ");
            actual = actual.next;
        }
        Console.WriteLine("None");
    }

    public bool Buscar(int data)
    {
        Nodo actual = head;
        while (actual != null)
        {
            if (actual.data == data) return true;
            actual = actual.next;
        }
        return false;
    }

    public void Eliminar(int data)
    {
        Nodo actual = head;
        Nodo previo = null;

        if (actual != null && actual.data == data)
        {
            head = actual.next;
            return;
        }

        while (actual != null && actual.data != data)
        {
            previo = actual;
            actual = actual.next;
        }

        if (actual == null) return;

        previo.next = actual.next;
    }

    public static void Main(string[] args)
    {
        ListaEnlazada lista = new ListaEnlazada();
        lista.InsertarAlFinal(10);
        lista.InsertarAlFinal(20);
        lista.InsertarAlFinal(30);
        Console.Write("Lista inicial: ");
        lista.ImprimirLista();

        lista.InsertarAlPrincipio(5);
        Console.Write("Lista después de insertar al principio: ");
        lista.ImprimirLista();

        Console.WriteLine("¿Esta el 20? " + lista.Buscar(20));
        Console.WriteLine("¿Esta el 99? " + lista.Buscar(99));

        lista.Eliminar(20);
        Console.Write("Lista después de eliminar 20: ");
        lista.ImprimirLista();

        lista.Eliminar(5);
        Console.Write("Lista después de eliminar 5: ");
        lista.ImprimirLista();
    }
}

File: LEM_Listas/Listas_Enlazadas.js
class Nodo {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class ListaEnlazada {
    constructor() {
        this.head = null;
    }

    insertar_al_principio(data) {
        const nuevo = new Nodo(data);
        nuevo.next = this.head;
        this.head = nuevo;
    }

    insertar_al_final(data) {
        const nuevo = new Nodo(data);
        if (!this.head) {
            this.head = nuevo;
            return;
        }
        let ultimo = this.head;
        while (ultimo.next) {
            ultimo = ultimo.next;
        }
        ultimo.next = nuevo;
    }

    imprimir_lista() {
        let actual = this.head;
        let output = "";
        while (actual) {
            output += actual.data + " -> ";
            actual = actual.next;
        }
        output += "None";
        console.log(output);
    }

    buscar(data) {
        let actual = this.head;
        while (actual) {
            if (actual.data === data) return true;
            actual = actual.next;
        }
        return false;
    }

    eliminar(data) {
        let actual = this.head;
        let previo = null;

        if (actual && actual.data === data) {
            this.head = actual.next;
            return;
        }

        while (actual && actual.data !== data) {
            previo = actual;
            actual = actual.next;
        }

        if (!actual) return;

        previo.next = actual.next;
    }
}

const lista = new ListaEnlazada();
lista.insertar_al_final(10);
lista.insertar_al_final(20);
lista.insertar_al_final(30);
process.stdout.write("Lista inicial: ");
lista.imprimir_lista();

lista.insertar_al_principio(5);
process.stdout.write("Lista después de insertar al principio: ");
lista.imprimir_lista();

console.log("¿Esta el 20? " + lista.buscar(20));
console.log("¿Esta el 99? " + lista.buscar(99));

lista.eliminar(20);
process.stdout.write("Lista después de eliminar 20: ");
lista.imprimir_lista();

lista.eliminar(5);
process.stdout.write("Lista después de eliminar 5: ");
lista.imprimir_lista();